code/batch_run_cli.sh

```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+# This script accepts a list of repository (dataset) IDs and their corresponding latest SHAs.
+# It runs the Neurobagel CLI on each dataset and updates the SHA in sha.txt if the CLI runs successfully.
+
+# A space-separated list where each item is: <repo ID>,<SHA>
+dataset_list_path=$1
+
+for dataset in $(cat $dataset_list_path); do
+    repo=$(echo $dataset | cut -d ',' -f 1)
+    sha=\"$(echo $dataset | cut -d ',' -f 2)\"
+    echo "Repo info: $repo,$sha"
+
+    echo "${repo}: Running the CLI"
+    ./run_cli_single_dataset.sh $repo
+
+    # If the CLI did not run successfully, exit loop without updating SHA
+    if [ $? -eq 0 ]; then
+        # Replace the old SHA with the new one
+        echo "${repo}: CLI ran successfully!"
+
+        line=$(grep "$repo" sha.txt)
+        # TODO: Check if we need this section?
+        # Prevents wf from exiting if the repo does not already exist in sha.txt (sed error)
+        if [ ! -z "$line" ]; then
+            echo "${repo}: Updating SHA in file"
+            sed -i "s/${line}/${repo},${sha}/" sha.txt
+        else
+            echo "${repo}: SHA not found, writing latest SHA to file"
+            echo $repo,$sha >> sha.txt
+        fi
```

Comment:
I think we don't need this section, since we're doing a check in the `bash_scraper` and `run_cli_on_all_repos` already collects all the SHAs.

Reply:
Bit tough to understand tbh. I'll try:

- `sha_scraper` is responsible for looking at all repos and ask "have I seen this repo before"
  - if yes: "has this repo changed since I was last here"
    - if yes: "could I do something about this repo (i.e. run CLI)"
      - if yes again: "add the repo to my todo list" **but not** update the repo sha in the lookup file
      - if no: **only** update the repo sha in the lookup file
  - if no (nothing has changed): do nothing
- if no ("never seen this repo"): -> go to "could I do something about this repo (i.e. run CLI)" and do that flow

So with that in mind, the section here seems to be the counterpart to "add the repo to my todo list" **but not** update the repo sha in the lookup file. Because I will **only** "update the repo sha in the lookup file" **if** my CLI run has succeeded. Otherwise I would not retry running the CLI next time I come by

Reply:
Yes, that's exactly it @surchs! Thanks for the clear summary.

I've further updated the context comments for sha_scraper.sh and batch_run_cli.sh to hopefully make this a bit clearer.

Essentially, in the current automation framework, the sha.txt is a file that tells us:
- The latest 'version' of each OpenNeuroDatasets-JSONLD repo that we were last able to either (A) successfully run the CLI on, **OR** (B) skipped running the CLI because they were missing required files

And that's because when a repo meets either of these conditions, we want to _not_ try to run the CLI again when the scheduled workflow next comes around, UNLESS there has been a change to the repo.

On the other hand, when a repo fails the CLI, for now we _do_ always want to try running the CLI again the next time the workflow comes around. And we trigger/achieve this by not updating the entries for those repos in sha.txt until they meet one of the previous two conditions.

These decisions are somewhat arbitrary, so please let me know / open an issue if you think any part of this doesn't make sense, and we can incrementally refine the workflow logic!

(Also, I'm going to add this to a hackmd for reference)